File: Code/Environments/AnalyseLightcurve.jl/src/AnalyseLightcurve.jl
module AnalyseLightcurve

using CairoMakie

#
## Constants
#

# Get the directory holding this script
const SCRIPT_DIR = dirname(abspath(@__FILE__))
# Get the JuliaWorkshop Path
const BASE_DIR = abspath(joinpath(SCRIPT_DIR, "../../../../"))
# Get the directory of the data
const DATA_DIR = joinpath(BASE_DIR, "Data")
# Get the data file
const DATA_FILE = joinpath(DATA_DIR, "data.csv")
# Define the output file
const OUTPUT_FILE = joinpath(SCRIPT_DIR, "output_jl.txt")
# Define the plot file
const PLOT_FILE = joinpath(SCRIPT_DIR, "Supernova Lightcurve")

"""
get_data()

    Read data from DATA_FILE into a dictionary
"""
function get_data()
    # Read file lines into an array
    raw_data = open(DATA_FILE, "r") do io
        return readlines(io)
    end

    # Headers are the first line
    headers = split(raw_data[1], ",")

    # Initialise data dictionary
    data = Dict(header => [] for header in headers)

    # Go row-by-row to read in data
    for row in raw_data[2:end]
        values = split(row, ",")
        for (col, header) in enumerate(headers)
            value = values[col]
            push!(data[header], value)
        end
    end

    return data
end

"""
analyse_data(data)

    Peform some simple data analysis, and write the result to a file
"""
function analyse_data(data)
    output = []

    # Get headers
    headers = collect(keys(data))
    push!(output, "Data has Headers: $headers")

    # Get number of elements
    num_elems = length(data[headers[1]])
    push!(output, "Data has $num_elems elements")

    output = join(output, "\n")
    open(OUTPUT_FILE, "w") do io
        write(io, output)
    end
end

"""
plot_data(data)

    Create a lightcurve plot of the data
"""
function plot_data(data)
    raw_bands = data["band"]
    raw_times = parse.(Float64, data["time"])
    raw_fluxes = parse.(Float64, data["flux"])
    raw_flux_errors = parse.(Float64, data["e_flux"])

    # Order of magnitude of the data
    oom = floor(log10(maximum(raw_fluxes)))

    bands = Set(raw_bands)
    masks = Dict(band => raw_bands .== band for band in bands)
    times = Dict(band => raw_times[masks[band]] for band in bands)
    fluxes = Dict(band => raw_fluxes[masks[band]] for band in bands)
    flux_errors = Dict(band => raw_flux_errors[masks[band]] for band in bands)

    f = Figure()
    ax = Axis(
        f[1, 1],
        xlabel="Time",
        ylabel="Flux",
        title="Supernova Lightcurves",
        yticks=MultiplesTicks(5, 10^oom, "e$oom"; strip_zero=true),
    )

    for band in bands
        time = times[band]
        flux = fluxes[band]
        flux_error = flux_errors[band]

        scatter_plot = scatter!(ax, time, flux, label=band)
        errorbars_plot = errorbars!(ax, time, flux, flux_error)

        # Special case for `K` band to make it less visible
        if band == "K"
            # Move to bottom of figure
            translate!(scatter_plot, 0, 0, -10)
            translate!(errorbars_plot, 0, 0, -10)
            # Shrink marker size
            scatter_plot.markersize = 0.5 * scatter_plot.markersize[]
            # Make colour transparent
            scatter_plot.color = (scatter_plot.color[], 0.5)
            # Shrink errorbar size
            errorbars_plot.linewidth = 0.5 * errorbars_plot.linewidth[]
            # Make colour transparent
            errorbars_plot.color = (errorbars_plot.color[], 0.5)
        end
    end

    Legend(f[1, 2], ax, "Bands")
    save("$PLOT_FILE.svg", f)
    save("$PLOT_FILE.png", f)
end

function main()
    @time begin
        data = get_data()
        analyse_data(data)
        plot_data(data)
    end
end

function (@main)(ARGS)
    main()
end

end # module AnalyseLightcurve

File: Code/Environments/AnalyseLightcurve.py/src/AnalyseLightcurve/AnalyseLightcurve.py
from os.path import realpath, dirname, join
from time import time

from matplotlib import pyplot as plt

#
## Constants
#

# Get the directory holding this script
SCRIPT_DIR = dirname(realpath(__file__))
# Get the JuliaWorkshop Path
BASE_DIR = realpath(join(SCRIPT_DIR, "../../../../../"))
# Get the directory of the data
DATA_DIR = join(BASE_DIR, "Data")
# Get the data file
DATA_FILE = join(DATA_DIR, "data.csv")
# Define the output file
OUTPUT_FILE = join(SCRIPT_DIR, "output_py.txt")
# Define the plot file
PLOT_FILE = join(SCRIPT_DIR, "Supernova Lightcurve")


def get_data():
    """
    get_data()

        Read data from DATA_FILE into a dictionary
    """

    # Read file lines into an array
    with open(DATA_FILE, "r") as io:
        raw_data = io.readlines()

    # Headers are the first line
    headers = raw_data[0].split(",")

    # Initialise data dictionary
    data = {header: [] for header in headers}

    # Go row-by-row to read in data
    for row in raw_data[1:]:
        values = row.split(",")
        for col, header in enumerate(headers):
            value = values[col]
            data[header].append(value)

    return data


def analyse_data(data):
    """
    analyse_data(data)

        Peform some simple data analysis, and write the result to a file
    """
    output = []

    # Get headers
    headers = list(data.keys())
    output.append(f"Data has Headers: {headers}")

    # Get number of elements
    num_elems = len(data[headers[0]])
    output.append(f"Data has {num_elems} elements")

    output = "\n".join(output)
    with open(OUTPUT_FILE, "w") as io:
        io.write(output)


def plot_data(data):
    """
    plot_data(data)

        Create a lightcurve plot of the data
    """
    raw_bands = data["band"]
    raw_times = [float(t) for t in data["time"]]
    raw_fluxes = [float(f) for f in data["flux"]]
    raw_flux_errors = [float(e) for e in data["e_flux"]]

    bands = set(raw_bands)
    masks = {band: [b == band for b in raw_bands] for band in bands}
    times = {
        band: [t for (i, t) in enumerate(raw_times) if masks[band][i]] for band in bands
    }
    fluxes = {
        band: [f for (i, f) in enumerate(raw_fluxes) if masks[band][i]]
        for band in bands
    }
    flux_errors = {
        band: [e for (i, e) in enumerate(raw_flux_errors) if masks[band][i]]
        for band in bands
    }

    f = plt.figure()
    ax = f.add_subplot(
        1, 1, 1, xlabel="Time", ylabel="Flux", title="Supernova Lightcurve"
    )

    for band in bands:
        time = times[band]
        flux = fluxes[band]
        flux_error = flux_errors[band]

        scatter_plot = ax.scatter(time, flux, label=band)
        color = scatter_plot.get_facecolor()
        errorbars_plot = ax.errorbar(
            time, flux, yerr=flux_error, fmt="none", ecolor=color
        )

        # Special case for `K` band to make it less visible
        if band == "K":
            # Move to bottom of figure
            scatter_plot.set_zorder(-10)
            # Shrink marker size
            scatter_plot.set_sizes([0.5 * size for size in scatter_plot.get_sizes()])
            # Make colour transparent
            scatter_plot.set_alpha(0.5)
            for child in errorbars_plot.get_children():
                # Move to bottom of figure
                child.set_zorder(-10)
                # Shrink errorbar size
                child.set_linewidth = 0.5 * child.get_linewidth()
                # Make colour transparent
                child.set_alpha(0.5)

    ax.legend(
        title="Bands",
        loc="center left",
        bbox_to_anchor=(1.0, 0.5),
    )
    f.savefig(f"{PLOT_FILE}.svg", bbox_inches="tight")
    f.savefig(f"{PLOT_FILE}.png", bbox_inches="tight")


def main():
    start = time()
    data = get_data()
    analyse_data(data)
    plot_data(data)
    end = time()
    print(f"script.py took {end - start} seconds")


if __name__ == "__main__":
    main()
